### PR TITLE
Allow for asynchronous subscription callbacks

### PR DESCRIPTION
**Pull request by chrisjbremner**
I am working a server implementation that would greatly benefit from being able to create subscriptions that have asynchronous callbacks. In this PR, I change the callback behavior for subscriptions so that you can use either synchronous or asynchronous definitions:

For example this would be valid:
```python
class MySubHandler:
    async def datachange_notification(self, node, val, data):
        await asyncio.sleep(1)
        print("got datachange")
    
    def event_notification(self, event):
        print("got event")
```

In addition, I made it so that the async callbacks run as tasks and the sync callbacks run in the event loop so that other processing is not blocked.

Would this make sense, or would you propose just using `asyncio.create_task(my_async_coro)` inside the callback instead?

### CODE REVIEW

**Comment by oroulet** [comment]
I thought this was already done. I think it is ok. It gives more flexibility to users. I would even telling the possibility for sync handlers

**Comment by oroulet** [comment]
@cbergmiller opinion?

**Comment by cbergmiller** [comment]
Looks like an improvement. I can't see any obvious problems in the changes.

File: asyncua/server/internal_subscription.py [hunk 1]
```diff
@@ -110,7 +110,10 @@ def publish_results(self):
             # The callback can be:
             #    Subscription.publish_callback -> server internal subscription
             #    UaProcessor.forward_publish_response -> client subscription
-            self.pub_result_callback(result)
+            if asyncio.iscoroutinefunction(self.pub_result_callback):
+                self.loop.create_task(self.pub_result_callback(result))
```
Comment (oroulet) [inline asyncua/server/internal_subscription.py:114]: ops I did not see that one. create tasks and forgetting them is not a good idea. Can't you make publish_callback() async? I do not think it is called form many places

Comment (chrisjbremner) [reply asyncua/server/internal_subscription.py:114]: `publish_callback` is already async, are you proposing I do `await self.pub_result_callback(result)` here? If so, we have to make `publish_results`. Making that change results in quite a long cascade of updates from sync functions to async, which wouldn't be that bad, but it eventually requires a change to `trigger_event` to async, which presumably would be a breaking change for anyone who relies on that being a sync function.

Comment (oroulet) [reply asyncua/server/internal_subscription.py:114]: There is another PR doing that in the pipeline anyway so maybe just go for it

**Comment by oroulet** [comment]
Looks good. But you need to rebase on master so we can merge

**Comment by oroulet** [comment]
github does not let me merge yet. try again to rebase on master `git pull -r origin master`

**Comment by oroulet** [comment]
nevermind I made it using command line

**Pull request closed without merging**